### DISCUSSION TITLE
Does Oceananigans support coastal modelling?

### DISCUSSION THREAD

**Discussion by stanleesocca** [CliMA/Oceananigans.jl#3270] [Ideas]
Hi devs. This is not an issue but more of comment. I will immediately close this after some feedbacks.

I was wondering if Oceananigans in its current form has support for coastal or regional modelling? I ask because my impression is that Oceananigans follows similar paradigm as MITgcm with the latter seemingly used in regional ocean modelling. Are there plans to use the model for such setup? What will it takes for the community to get to such level?

Thanks.

**Comment by simone-silvestri** [comment]
Hello @stanleesocca,

at the moment Oceananigans can be used for regional modeling with some hacks, but we have plans to make the whole process more streamlined.
In particular, we are planning to implement a utility that allows easy handling of time-dependent boundary conditions and inlets/outlets, and we still need to implement an easy way to deal with atmosphere-ocean interaction through bulk formulae.
In addition, we still do not support tidal parameterization or wetting/drying which might be required for regional modeling.

What setup were you thinking to implement?

**Reply by glwagner** [reply]
"Coastal modeling" can mean a lot of different things (ranging from idealized to realistic with complex bathymetry), so we need more specific information to provide a useful answer. The short answer is that Oceananigans _can_ be used for "coastal modeling", broadly speaking, but depending on the application more user input may be required than if using MITgcm, because some convenience features (some of which are mentioned by @simone-silvestri) are missing.

The more involved answer is that some numerical methods in MITgcm that prove useful for coastal modeling (for example things that are important for tides in shallow areas like z-star vertical coordinate, nonlinear free surface, mature partial cell implementation) are not implemented in Oceananigans. Bulk formula are easy to code up, but numerical methods for strong tidal flows would be more difficult.

**Reply by stanleesocca** [reply]
Thanks for the reply @simone-silvestri @glwagner. My setup (though not urgently needed) is something akin to a realistic simulation with complex bathymetric layout @glwagner . I understand Oceananigans isn't yet ripe for such but I am excited and curious about such possibility in the future. 
In addition to your comments, I suspect that the vertical grid used in Oceananigans doesn't support sigma coordinate system (or do but I'm not aware of) like some other modelling platform do. Also, I wonder how the ocean-boundary condition will be implemented in such case with Oceananigans @simone-silvestri

**Reply by glwagner** [reply]
Correct -- we only support z grid right now. However, bathymetry and surface boundary conditions can be as complex as you like.

**Reply by stanleesocca** [reply]
Cool. Aside the surface boundary condition. From the docs, it isn't clear to me how one can include a bathymetry grid file. Any pointer?

**Reply by glwagner** [reply]
For simulations in complex domains, we use `ImmersedBoundaryGrid`. You're right that it isn't documented --- that should be fixed.

Here is an example script that uses `ImmersedBoundaryGrid`. That same directly has a few more. Perhaps @simone-silvestri has more suggestions:

https://github.com/CliMA/Oceananigans.jl/blob/main/validation/immersed_boundaries/bumpy_baroclinic_adjustment.jl

**Reply by stanleesocca** [reply]
Thanks @glwagner. Really appreciate. I will take a look and get back to you if I run into any trouble.

**Reply by glwagner** [reply]
This PR is also adding an example (I forgot about this):

https://github.com/CliMA/Oceananigans.jl/pull/3132

motivation to merge that sooner.
